src/libs/pluginprocess/pluginprocess.c

```diff
@@ -0,0 +1,379 @@
+ *
+ * Executes plugins in a separate process via fork and uses a simple
+ * communication protocol based on the dump plugin via named pipes.
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include "kdbpluginprocess.h"
+#include <kdbinvoke.h>
+// To access the plugin function pointers
+#include <kdberrors.h>
+#include <kdblogger.h>
+#include <kdbprivate.h>
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+struct _ElektraPluginProcess
+{
+	char * commandPipe;
+	char * resultPipe;
+	int pid;
+	int counter;
+	ElektraInvokeHandle * dump;
+};
+
+static void cleanupPluginData (ElektraPluginProcess * pp)
+{
+	elektraInvokeClose (pp->dump);
+	if (pp->pid) unlink (pp->commandPipe);
+	elektraFree (pp->commandPipe);
+	if (pp->pid) unlink (pp->resultPipe);
+	elektraFree (pp->resultPipe);
+	elektraFree (pp);
+}
+
+static char * intToStr (int i)
+{
+	int size;
+	char * str;
+	size = snprintf (NULL, 0, "%d", i);
+	str = elektraMalloc (size + 1);
+	size = snprintf (str, size + 1, "%d", i);
+	return str;
+}
+
+/** Start the child process' command loop
+ *
+ * This will make the child process wait for plugin commands
+ * and execute them, returning the result to the parent. This
+ * is typically called in a plugin's open function.
+ *
+ * @param handle the plugin's handle
+ * @param pp the data structure containing the plugin's process information
+ * @see elektraPluginProcessInit how to use this function in a plugin
+ * @ingroup processplugin
+ **/
+void elektraPluginProcessStart (Plugin * handle, ElektraPluginProcess * pp)
+{
+	Key * commandPipeKey = keyNew ("/pluginprocess/commandPipe", KEY_VALUE, pp->commandPipe, KEY_END);
+	Key * resultPipeKey = keyNew ("/pluginprocess/resultPipe", KEY_VALUE, pp->resultPipe, KEY_END);
+	KeySet * keySet = ksNew (0, KS_END);
+	int counter = 0;
+	plugin_t command = ELEKTRA_PLUGIN_END;
+
+	do
+	{
+		ELEKTRA_LOG_DEBUG ("C: Wait for a KeySet");
+		elektraInvoke2Args (pp->dump, "get", keySet, commandPipeKey);
+		ELEKTRA_LOG_DEBUG ("C: We received a KeySet with %zd keys in it", ksGetSize (keySet));
+		Key * commandKey = ksLookupByName (keySet, "/pluginprocess/command", KDB_O_POP);
+		char * endPtr;
+		// We'll always write some int value into it, so this should be fine
+		int prevErrno = errno;
+		errno = 0;
+		command = (int)strtol (keyString (commandKey), &endPtr, 10);
+
+		Key * originalNameKey = ksLookupByName (keySet, "/pluginprocess/original", KDB_O_POP);
+		Key * originalKey = ksLookupByName (keySet, keyString (originalNameKey), KDB_O_NONE);
+
+		int result = ELEKTRA_PLUGIN_STATUS_ERROR;
+		if (*endPtr == '\0' && errno != ERANGE)
+		{
+			ELEKTRA_LOG ("C: We want to execute the command with the value %d now", command);
+			switch (command)
+			{
+			case ELEKTRA_PLUGIN_OPEN:
+				counter++;
+				result = handle->kdbOpen (handle, originalKey);
```

Comment:
Why calling functions of the plugin itself, isn't this a proxy plugin to call other plugins?

Reply:
No i think we have a misunderstanding here. It is a library you can use to add the forking and communication functionality to your plugins, not a plugin on its own. [I'll link the usage example again and i hope this clears it up.](https://github.com/e1528532/libelektra/blob/haskell-pluginprocess/src/plugins/haskell/haskell.c.in).

I was not really sure how to design this as a plugin on its own, upon mounting people most like won't like to have to deal with that and simply mount a specification e.g. via `kdb mount /testSpecification.ini /spec/test ini typechecker` (assuming typechecker is a haskell plugin using this library) and not something like `kdb mount .... ini processplugin proxy=typechecker` (using a configuration to specify which plugin shall be proxied).

Maybe there is a simple way to have it as a plugin though, in that case I'd be fine with refactoring it to a plugin as the general logic most likely will stay pretty much the same.

Reply:
Ahh, okay. I see. Yes, it is a good idea to have it as library. Then plugins which do not work at all without being in a separate process can link against it.

For those who both modes make somehow sense (e.g. crypto) we should implement the plugin nevertheless. With your library writing such a plugin would be trivial anyway.

The usage would be `kdb mount file.ini ini process plugin=crypto` to have the crypto plugin running in a separate process. Let us discuss this today.

Reply:
Ok so this should be fine in the context that it's a library, a plugin that uses this library to implement the functionality you described will follow.

Reply:
Thank you! It is easier to comment if you create PRs, even if they are far from finished.

When looking at your code, I was missing a proxy for the contract: For the clause "infos/placements" you need to return what the plugin we are a proxy for returned. (You will need this logic also for the haskell plugin.). You can achieve this by renaming the respective clauses of the original plugin.

If you have any troubles implementing, please create the PR then we can easier discuss the problems.